3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/SendgridAdapter.scala

```diff
@@ -17,105 +17,128 @@ package common
+    "deferred" -> SchemaKey("com.sendgrid", "deferred", "jsonschema", "1-0-0").toSchemaUri,
+    "bounce" -> SchemaKey("com.sendgrid", "bounce", "jsonschema", "1-0-0").toSchemaUri,
+    "open" -> SchemaKey("com.sendgrid", "open", "jsonschema", "1-0-0").toSchemaUri,
+    "click" -> SchemaKey("com.sendgrid", "click", "jsonschema", "1-0-0").toSchemaUri,
+    "spamreport" -> SchemaKey("com.sendgrid", "spamreport", "jsonschema", "1-0-0").toSchemaUri,
+    "unsubscribe" -> SchemaKey("com.sendgrid", "unsubscribe", "jsonschema", "1-0-0").toSchemaUri,
+    "group_unsubscribe" -> SchemaKey("com.sendgrid", "group_unsubscribe", "jsonschema", "1-0-0").toSchemaUri,
+    "group_resubscribe" -> SchemaKey("com.sendgrid", "group_resubscribe", "jsonschema", "1-0-0").toSchemaUri
   )
 
+
   /**
-   * Converts a CollectorPayload instance into raw events.
-   * A Sendgrid Tracking payload only contains a single event.
-   * We expect the name parameter to be 1 of 6 options otherwise
-   * we have an unsupported event type.
-   *
-   * @param payload The CollectorPayload containing one or more
-   *        raw events as collected by a Snowplow collector
-   * @param resolver (implicit) The Iglu resolver used for
-   *        schema lookup and validation. Not used
-   * @return a Validation boxing either a NEL of RawEvents on
-   *         Success, or a NEL of Failure Strings
-   */
+    *
+    * Converts a payload into a list of validated events
+    * Expects a valid json - returns a single failure if one is not present
+    *
+    * @param body json payload as POST'd by sendgrid
+    * @param payload the rest of the payload details
+    * @return a list of validated events, successes will be the corresponding raw events
+    *         failures will contain a non empty list of the reason(s) for the particular event failing
+    */
+  private def payloadBodyToEvents(body: String, payload: CollectorPayload): List[Validated[RawEvent]] = {
+    try {
+      val parsed = parse(body)
+
+      if (parsed.children.isEmpty) {
+        return List(s"$VendorName event failed json sanity check: has no events".failNel)
+      }
+
+      for ((itm, index) <- parsed.children.zipWithIndex)
+        yield {
+          val eventType = (itm \\ "event").extractOpt[String]
+          val queryString = toMap(payload.querystring)
+
+          lookupSchema(eventType, VendorName, index, EventSchemaMap) match {
+            case Success(schema) => {
+              Success(
+                RawEvent(
+                  api = payload.api,
+                  parameters = toUnstructEventParams(TrackerVersion,
+                    queryString,
+                    schema,
+                    itm,
+                    "srv"),
+                  contentType = payload.contentType,
+                  source = payload.source,
+                  context = payload.context
+                )
+              )
+            }
+            case Failure(err) => Failure(err)
+          }
+        }
+
```

Comment:
No problem! Wanting to do something to a Validation if it's a Success but leave it unchanged if it's a Failure is pretty common case.